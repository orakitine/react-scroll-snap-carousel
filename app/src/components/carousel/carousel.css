@import './components/carousel-base/carousel-base.css';
@import './components/carousel-slides/carousel-slides.css';
@import './components/carousel-arrow/carousel-arrow.css';
@import './components/carousel-pagination/carousel-pagination.css';
