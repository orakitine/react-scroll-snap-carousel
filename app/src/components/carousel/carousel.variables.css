@import '../../variables.css';

:root {
  --carousel-slides-spacing-sm: var(--spacing-01);
  --carousel-slides-spacing-md: var(--spacing-02);
  --carousel-slides-spacing-lg: var(--spacing-03);
  --carousel-slides-spacing-xl: var(--spacing-04);
  --carousel-pagination-dot-size: 6px;
  --carousel-pagination-dot-padding: 4px;
  --carousel-arrow-padding: 0.25rem;
  --carousel-scroll-hide-spacing: 10px;
}
