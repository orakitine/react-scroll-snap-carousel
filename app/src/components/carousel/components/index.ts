export * from './carousel-base/carousel-base.component';
export * from './carousel-base/carousel-base.types';
export * from './carousel-arrow/carousel-arrow.component';
export * from './carousel-arrow/carousel-arrow.types';
export * from './carousel-pagination/carousel-pagination.component';
export * from './carousel-pagination/carousel-pagination.types';
export * from './carousel-slides/carousel-slides.component';
export * from './carousel-slides/carousel-slides.types';
