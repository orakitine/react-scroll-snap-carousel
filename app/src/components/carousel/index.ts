export * from './carousel.component';
export * from './carousel.types';
export * from './types';
export * from './components/carousel-base/carousel-base.types';
