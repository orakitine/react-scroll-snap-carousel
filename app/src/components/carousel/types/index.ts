export * from './carousel-arrow-increment.types';
export * from './carousel-indicator.types';
export * from './break-point-props.types';

