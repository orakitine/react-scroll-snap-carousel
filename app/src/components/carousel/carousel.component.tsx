import * as React from 'react';
import { debounce } from 'throttle-debounce';

import { BreakpointKey, getCurrentMediaQueryBreakpoint, isInBrowser } from '../../utils';

import { CarouselProps } from './carousel.types';
import { CarouselBase } from './components/carousel-base/carousel-base.component';
import { ARROW_INCREMENT_PAGE, INDICATOR_PAGES } from './types';

interface CarouselState {
  breakpoint: BreakpointKey;
}

export class Carousel extends React.Component<CarouselProps, CarouselState> {
  private onWindowSize = debounce(100, (_: Event) => {
    this.setState({
      breakpoint: getCurrentMediaQueryBreakpoint(),
    });
  });

  constructor(props: CarouselProps) {
    super(props);

    this.state = {
      breakpoint: getCurrentMediaQueryBreakpoint(),
    };
  }

  componentDidMount = () => {
    if (isInBrowser()) {
      window.addEventListener('resize', this.onWindowSize);
    }
  };

  componentWillUnmount = () => {
    if (isInBrowser()) {
      window.removeEventListener('resize', this.onWindowSize);
    }
  };

  render() {
    const {
      displayCount = 1,
      indicator = INDICATOR_PAGES,
      index = 0,
      children,
      arrowsIncrement = ARROW_INCREMENT_PAGE,
      noPadding = false,
      ...rest
    } = this.props;

    const displayCountProp = Math.max(
      typeof displayCount === 'number' ? displayCount : displayCount[this.state.breakpoint],
      1
    );
    const slidesPerPage = Math.floor(displayCountProp);

    const indicatorProp = typeof indicator === 'string' ? indicator : indicator[this.state.breakpoint];
    const arrowsIncrementProp =
      typeof arrowsIncrement === 'string' ? arrowsIncrement : arrowsIncrement[this.state.breakpoint];
    const slidesTotal = React.Children.toArray(children).length;
    const maxIndex = Math.max(slidesTotal - slidesPerPage, 0);
    const indexProp = index < 0 ? 0 : Math.min(maxIndex, index);
    const noPaddingProp = noPadding || displayCountProp === 1;

    return (
      <CarouselBase
        displayCount={displayCountProp}
        indicator={indicatorProp}
        index={indexProp}
        maxIndex={maxIndex}
        slidesTotal={slidesTotal}
        arrowIncrement={arrowsIncrementProp}
        slidesPerPage={slidesPerPage}
        noPadding={noPaddingProp}
        {...rest}
      >
        {children}
      </CarouselBase>
    );
  }
}
