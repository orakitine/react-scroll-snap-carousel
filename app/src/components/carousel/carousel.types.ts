import { BreakPointProp, CarouselArrowIncrement, CarouselIndicator } from "./types";

export interface CarouselProps extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * Number of Slides to show in carousel. Could be a decimal value. Minimum is 1
   * @default 1
   */
  displayCount?: BreakPointProp<number> | number;

  /**
   * Style of the carousel position indicator
   * 'pages' - Displays clickable bullet points ("meatballs")
   * 'number' - Ex. 2/10
   * 'none' - nothing
   * @default pages
   */
  indicator?: BreakPointProp<CarouselIndicator> | CarouselIndicator;

  /**
   * Number of slides to scroll by for side arrow buttons
   * 'none' - Arrow will not be rendered
   * 'single' - Arrows moves one slide over
   * 'page'; - Arrow moves whole amount of displayed items.
   * @default page
   */
  arrowsIncrement?: BreakPointProp<CarouselArrowIncrement> | CarouselArrowIncrement;

  /**
   * Contents for the slides.
   * @required
   */
  children: React.ReactNode;

  /**
   * Sets current index of the carousel. It is 0 based. Indexed item will be first slide on the left.
   * @default 0
   */
  index?: number;

  /**
   * A placeholder Element for slides that are not rendered.
   * If it is present, there will be limited number of slides rendered.
   * Rendered slides are Math.max(5, Math.ceil(displayCount) * 2) on each side of the visible portion of the carousel.
   */
  slidePlaceholder?: JSX.Element;

  /**
   * A query string to look up an element in the slide.
   * If it is present, navigation arrow will be centered relative to that element.
   * If string is not provided or element is not found Arrow will be vertically centered
   * Must be a valid CSS selector string.
   */
  arrowPositionSelector?: string;

  /**
   * Removes padding from slides if true.
   * @default false
   */
  noPadding?: boolean;

  /**
   * An event that is fired every time Carousel successfully changes index.
   */
  onIndexChange?: (index: number) => void;

  /**
   * Text string for the Skip carousel content button. It will show up when focused. Ex: 'Skip carousel content'
   * @required
   */
  skipLinkText: string;

  /**
   * Optional id for a skip link to go to. If provided Skip button will be an anchor link to this id.
   */
  skipToId?: string;

  /**
   * Text string to describe Carousel and its contents
   * @required
   */
  carouselAriaLabel: string;
}
