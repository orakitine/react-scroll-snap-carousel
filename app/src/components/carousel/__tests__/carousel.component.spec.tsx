import * as React from "react";
import testComponentHelper from "../../../../../test/unit/helpers/ComponentHelper";
import { Carousel } from "../carousel.component";

describe("<Header />", () => {
  const initialProps = {
    displayCount: 2.2,
    skipLinkText: "Skip carousel content",
    carouselAriaLabel: "A carousel of cats",
    slidePlaceholder: <div>placeholder</div>,
    children: (
      <React.Fragment>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
        <div>5</div>
        <div>6</div>
      </React.Fragment>
    )
  };
  const renderComponent = testComponentHelper(Carousel);

  describe("@renders", () => {
    it("in default state", () => {
      expect(renderComponent().getHtml()).toMatchSnapshot();
    });
  });
});
