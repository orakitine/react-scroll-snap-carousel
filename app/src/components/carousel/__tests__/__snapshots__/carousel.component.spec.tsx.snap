// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> @renders in default state 1`] = `
<Carousel>
  <CarouselBase
    arrowIncrement="page"
    displayCount={1}
    index={0}
    indicator="pages"
    maxIndex={1}
    noPadding={true}
    slidesPerPage={1}
    slidesTotal={2}
  >
    <Measure
      onResize={[Function]}
    >
      <Component
        contentRect={
          Object {
            "bounds": Object {},
            "client": Object {},
            "entry": Object {},
            "margin": Object {},
            "offset": Object {},
            "scroll": Object {},
          }
        }
        measure={[Function]}
        measureRef={[Function]}
      >
        <div
          className="carousel"
          index={0}
        >
          <button
            className="carouselSkip"
            onClick={[Function]}
          />
          <div
            className="carouselScrollbarHide"
          >
            <div
              className="carouselContainer"
            >
              <ul
                className="carouselFrame"
                onScroll={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
              >
                <CarouselSlides
                  currentIndex={0}
                  displayCount={1}
                  noPadding={true}
                  slideWidth={0}
                >
                  <li
                    className="carouselSlide"
                    key="0"
                    style={
                      Object {
                        "width": "0px",
                      }
                    }
                  >
                    <div
                      className="carouselSlideContainer"
                    >
                       
                    </div>
                  </li>
                  <li
                    className="carouselSlide"
                    key="1"
                    style={
                      Object {
                        "width": "0px",
                      }
                    }
                  >
                    <div
                      className="carouselSlideContainer"
                    >
                       
                    </div>
                  </li>
                </CarouselSlides>
              </ul>
              <CarouselArrow
                direction="next"
                onClick={[Function]}
              >
                <button
                  aria-hidden={true}
                  className="carouselArrow carouselArrowRight"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <Icon
                    name="chevron-right"
                  >
                    <svg
                      viewBox="0 0 21 21.125"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12.475 10.6l-7.3 7.3a2 2 0 0 0 0 2.7 2 2 0 0 0 2.7 0l8.6-8.7a2 2 0 0 0 0-2.7L7.875.6a1.936 1.936 0 0 0-1.3-.6 1.612 1.612 0 0 0-1.3.6 2 2 0 0 0 0 2.7z"
                      />
                    </svg>
                  </Icon>
                </button>
              </CarouselArrow>
            </div>
          </div>
          <CarouselPagination
            currentIndex={0}
            forceLast={false}
            indicatorType="pages"
            itemsPerPage={1}
            onPageClick={[Function]}
            total={2}
          >
            <div
              className="carouselPagination"
            >
              <button
                aria-hidden={true}
                className="carouselPageCurrent"
                key="0"
                onClick={[Function]}
                tabIndex={-1}
              />
              <button
                aria-hidden={true}
                className="carouselPage"
                key="1"
                onClick={[Function]}
                tabIndex={-1}
              />
            </div>
          </CarouselPagination>
          <div
            aria-hidden="true"
            tabIndex={-1}
          />
        </div>
      </Component>
    </Measure>
  </CarouselBase>
</Carousel>
`;
